packages/nifti-volume-loader/src/helpers/fetchAndAllocateNiftiVolume.ts

```diff
@@ -0,0 +1,266 @@
+import scaleNiftiArray from './scaleNiftiArray';
+import Events from '../enums/Events';
+import { NIFTI_LOADER_SCHEME } from '../constants';
+
+const { createUint8SharedArray, createFloat32SharedArray } = utilities;
+
+export const urlsMap = new Map();
+
+function fetchArrayBuffer(url, onProgress, signal, onload) {
+  return new Promise((resolve, reject) => {
+    const xhr = new XMLHttpRequest();
+    xhr.open('GET', url, true);
+    xhr.responseType = 'arraybuffer';
+
+    const onLoadHandler = function (e) {
+      if (onload && typeof onload === 'function') {
+        onload();
+      }
+
+      // Remove event listener for 'abort'
+      if (signal) {
+        signal.removeEventListener('abort', onAbortHandler);
+      }
+
+      resolve(xhr.response);
+    };
+
+    const onAbortHandler = () => {
+      xhr.abort();
+
+      // Remove event listener for 'load'
+      xhr.removeEventListener('load', onLoadHandler);
+
+      reject(new Error('Request aborted'));
+    };
+
+    xhr.addEventListener('load', onLoadHandler);
+
+    if (onProgress && typeof onProgress === 'function') {
+      xhr.onprogress = function (e) {
+        onProgress(e.loaded, e.total);
+      };
+    }
+
+    if (signal && signal.aborted) {
+      xhr.abort();
+      reject(new Error('Request aborted'));
+    } else if (signal) {
+      signal.addEventListener('abort', onAbortHandler);
+    }
+
+    xhr.send();
+  });
+}
+
+export const getTypedNiftiArray = (datatypeCode, niftiImageBuffer) => {
+  switch (datatypeCode) {
+    case NIFTICONSTANTS.NIFTI_TYPE_UINT8:
+      return new Uint8Array(niftiImageBuffer);
+    case NIFTICONSTANTS.NIFTI_TYPE_FLOAT32:
+      return new Float32Array(niftiImageBuffer);
+    case NIFTICONSTANTS.NIFTI_TYPE_INT16:
+      return new Int16Array(niftiImageBuffer);
+    default:
+      throw new Error(`datatypeCode ${datatypeCode} is not yet supported`);
+  }
+};
+
+export default async function fetchAndAllocateNiftiVolume(
+  volumeId: string
+): Promise<NiftiImageVolume> {
+  // nifti volumeIds start with 'nifti:' so we need to remove that
+  const niftiURL = volumeId.substring(NIFTI_LOADER_SCHEME.length + 1);
+
+  const progress = (loaded, total) => {
+    const data = { volumeId, loaded, total };
+    triggerEvent(eventTarget, Events.NIFTI_VOLUME_PROGRESS, {
+      data,
+    });
+  };
+
+  const onLoad = () => {
+    const data = { volumeId };
+    triggerEvent(eventTarget, Events.NIFTI_VOLUME_LOADED, {
+      data,
+    });
+  };
+
+  const controller = new AbortController();
+  const signal = controller.signal;
+
+  urlsMap.set(niftiURL, { controller, loading: true });
+
+  let niftiBuffer = (await fetchArrayBuffer(
+    niftiURL,
+    progress,
+    signal,
+    onLoad
+  )) as ArrayBuffer;
+
+  urlsMap.delete(niftiURL);
+
+  let niftiHeader = null;
+  let niftiImage = null;
+
+  if (NiftiReader.isCompressed(niftiBuffer)) {
+    niftiBuffer = NiftiReader.decompress(niftiBuffer);
+  }
+
+  if (NiftiReader.isNIFTI(niftiBuffer)) {
```

Comment:
This will throw a NPE if it isn't NIFTI data - probably throw an error if it isn't nifti data specifically, and otherwise continue handling it.